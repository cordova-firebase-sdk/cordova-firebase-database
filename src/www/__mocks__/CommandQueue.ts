export const execCmd = jest.fn((params: any): Promise<any> => {
  return Promise.resolve(params);
});
