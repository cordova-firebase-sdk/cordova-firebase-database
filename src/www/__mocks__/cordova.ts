export const exec = jest.fn((onSuccess, onError, pluginName, methodName, args): any => {
  setTimeout(onSuccess, 1);
  return {
    args,
    methodName,
    pluginName,
  };
});
