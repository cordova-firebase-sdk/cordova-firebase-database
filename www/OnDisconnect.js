


var BaseArrayClass = require('./BaseArrayClass'),
  execCmd = require('./FirebaseDatabaseCommandQueue'),
  LZString = require('./LZString');

/*******************************************************************************
 * @name OnDisconnect
 ******************************************************************************/
function OnDisconnect(pluginName) {
  var self = this,
    cmdQueue = new BaseArrayClass();

  Object.defineProperty(self, '_isReady', {
    value: false,
    writable: true
  });
  Object.defineProperty(self, 'id', {
    value: this.hashCode + '_onDisconnect'
  });
  Object.defineProperty(self, 'pluginName', {
    value: pluginName
  });

  cmdQueue._on('insert_at', function() {
    if (!self._isReady) return;

    var cmd;
    while(cmdQueue.getLength() > 0) {
      cmd = cmdQueue.removeAt(0, true);
      if (cmd && cmd.target && cmd.args) {
        execCmd.apply(cmd.target, cmd.args);
      }
    }
  });

  Object.defineProperty(self, '_cmdQueue', {
    value: cmdQueue,
    writable: false,
    enumerable: false
  });

}

OnDisconnect.prototype._privateInit = function() {
  this._isReady = true;
  this._cmdQueue._trigger('insert_at');
};

OnDisconnect.prototype._exec = function() {
  this._cmdQueue.push.call(this._cmdQueue, {
    target: this,
    args: Array.prototype.slice.call(arguments, 0)
  });
};


//---------------------------------------------------------------------------------
// onDisconnect.cancel
// https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#cancel
//---------------------------------------------------------------------------------
OnDisconnect.prototype.cancel = function(onComplete) {
  var self = this;

  return new new Promise(function(resolve, reject) {
    self._exec(function() {
      resolve.call(self);
      if (typeof onComplete === 'function') {
        onComplete.call(self);
      }
    }, function(error) {
      reject.call(self, error);
      if (typeof onComplete === 'function') {
        onComplete.call(self, error);
      }
    }, self.pluginName, 'onDisconnect_cancel', [{
      targetId: self.id
    }]);
  });
};



//---------------------------------------------------------------------------------
// onDisconnect.remove
// https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#remove
//---------------------------------------------------------------------------------
OnDisconnect.prototype.remove = function(onComplete) {
  var self = this;

  return new new Promise(function(resolve, reject) {
    self._exec(function() {
      resolve.call(self);
      if (typeof onComplete === 'function') {
        onComplete.call(self);
      }
    }, function(error) {
      reject.call(self, error);
      if (typeof onComplete === 'function') {
        onComplete.call(self, error);
      }
    }, self.pluginName, 'onDisconnect_remove', [{
      targetId: self.id
    }]);
  });
};


//---------------------------------------------------------------------------------
// onDisconnect.set
// https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#set
//---------------------------------------------------------------------------------
OnDisconnect.prototype.set = function(value, onComplete) {
  var self = this;

  return new new Promise(function(resolve, reject) {
    self._exec(function() {
      resolve.call(self);
      if (typeof onComplete === 'function') {
        onComplete.call(self);
      }
    }, function(error) {
      reject.call(self, error);
      if (typeof onComplete === 'function') {
        onComplete.call(self, error);
      }
    }, self.pluginName, 'onDisconnect_set', [{
      targetId: self.id,
      value: LZString.compressToBase64(JSON.stringify(value))
    }]);
  });
};



//---------------------------------------------------------------------------------
// onDisconnect.setWithPriority
// https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#setWithPriority
//---------------------------------------------------------------------------------
OnDisconnect.prototype.setWithPriority = function(value, priority, onComplete) {
  var self = this;

  return new new Promise(function(resolve, reject) {
    self._exec(function() {
      resolve.call(self);
      if (typeof onComplete === 'function') {
        onComplete.call(self);
      }
    }, function(error) {
      reject.call(self, error);
      if (typeof onComplete === 'function') {
        onComplete.call(self, error);
      }
    }, self.pluginName, 'onDisconnect_setWithPriority', [{
      targetId: self.id,
      value: LZString.compressToBase64(JSON.stringify(value)),
      priority: priority
    }]);
  });
};



//---------------------------------------------------------------------------------
// onDisconnect.update
// https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#update
//---------------------------------------------------------------------------------
OnDisconnect.prototype.update = function(values, onComplete) {
  var self = this;

  return new new Promise(function(resolve, reject) {
    self._exec(function() {
      resolve.call(self);
      if (typeof onComplete === 'function') {
        onComplete.call(self);
      }
    }, function(error) {
      reject.call(self, error);
      if (typeof onComplete === 'function') {
        onComplete.call(self, error);
      }
    }, self.pluginName, 'onDisconnect_update', [{
      targetId: self.id,
      values: LZString.compressToBase64(JSON.stringify(values))
    }]);
  });
};



module.exports = OnDisconnect;
