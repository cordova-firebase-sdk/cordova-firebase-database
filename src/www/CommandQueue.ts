import { exec } from "cordova";
// import { Promise } from "es6-promise";
import { nextTick, PluginBase } from "cordova-firebase-core/index";

declare let Promise: any;

/**
 * Parameters for execCmd function
 */
export interface IExecCmdParams {

  /**
   * Plugin's context
   */
  context: PluginBase;

  /**
   * Plugin's name in native side.
   * If omit this, context.id is used.
   */
  pluginName?: string;

  /**
   * Execute method name
   */
  methodName: string;

  /**
   * parameters for native side
   */
  args?: Array<any>;

  /**
   * synchronize options
   */
  execOptions?: {
    sync?: boolean;
    remove?: boolean;
  };

  [key: string]: any;
}

const commandQueue: Array<any> = [];
let isWaitMethod: string;
let isExecuting: boolean;
let executingCnt: number = 0;
const MAX_EXECUTE_CNT: number = 10;

let stopRequested: boolean = false;
window.addEventListener("unload", (): void => {
  stopRequested = true;
}, {
  once: true,
});

export const execCmd = (params: IExecCmdParams): Promise<any> => {
  params.execOptions = params.execOptions || {};
  params.args = params.args || [];
  params.pluginName = params.pluginName || params.context.id;

  // If the instance has been removed, do not execute any methods on it
  // except remove function itself.
  if (params.context.isRemoved && params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because removed.");
    return Promise.resolve();
  }

  // If the overlay is not ready in native side,
  // do not execute any methods except remove on it.
  if (!params.context.isReady && !params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because it's not ready.");
    return Promise.resolve();
  }

  return new Promise((resolve: (...params?: Array<any>) => void, reject: (...params?: Array<any>) => void) => {
    commandQueue.push({
      request: params,

      onSuccess: (...results: Array<any>): void => {
        // -------------------------------
        // success callback
        // -------------------------------

        // Even if the method was successful,
        // but the _stopRequested flag is true,
        // do not execute further code.
        if (!stopRequested) {
          resolve.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

      onError: (...results: Array<any>): void => {
        // -------------------------------
        // error callback
        // -------------------------------
        if (!stopRequested) {
          reject.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

    });

    // In order to execute all methods in safe,
    // the maps plugin limits the number of execution in a moment to 10.
    //
    // Note that Cordova-Android drops has also another internal queue,
    // and the internal queue drops our statement if the app send too much.
    //
    // Also executing too much statements at the same time,
    // it would cause many errors in native side, such as out-of-memory.
    //
    // In order to prevent these troubles, the maps plugin limits the number of execution is 10.
    if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
      return;
    }

    privateExec();
  });
};


const privateExec = (): void => {

  // You probably wonder why there is this code because it's already simular code at the end of the execCmd function.
  //
  // Because the commandQueue might change after the last code of the execCmd.
  // (And yes, it was occurred.)
  // In order to block surely, block the execution again.
  if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
    return;
  }

  isExecuting = true;

  // Execute some execution requests (up to 10) from the commandQueue
  let task: any;
  while (commandQueue.length > 0 && executingCnt < MAX_EXECUTE_CNT) {
    if (!stopRequested) {
      executingCnt++;
    }

    // Pick up the head one.
    task = commandQueue.shift();

    // push out normal tasks if stopRequested becomes true
    if (stopRequested && !task.request.execOptions.remove) {
      executingCnt--;
      continue;
    }

    // Some methods have to block other execution requests.
    if (task.request.execOptions.sync) {
      isWaitMethod = task.request.methodName;
      // console.log(`[sync start] ${task.request.pluginName}.${task.request.methodName}`);
      exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
      break;
    }

    // console.log(`[start] ${task.request.pluginName}.${task.request.methodName}`);
    exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
  }

  isExecuting = false;
};
