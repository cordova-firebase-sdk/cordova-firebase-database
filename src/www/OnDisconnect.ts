import { BaseArrayClass, LZString, PluginBase } from "cordova-firebase-core/index";
import { execCmd, IExecCmdParams } from "./CommandQueue";

export class OnDisconnect extends PluginBase {

  private _queue: BaseArrayClass = new BaseArrayClass();

  private _pluginName: string;

  constructor(pluginName: string) {
    super("OnDisconnect");




    this._pluginName = pluginName;

    this._queue._on("insert_at", (): void => {
      if (!this._isReady) {
        return;
      }
      if (this._queue._getLength() > 0) {
        const cmd: any = this._queue._removeAt(0, true);
        if (cmd && cmd.context && cmd.methodName) {
          execCmd(cmd).then(cmd.resolve).catch(cmd.reject);
        }
      }
      if (this._queue._getLength() > 0) {
        this._queue._trigger("insert_at");
      }
    });

  }

  public get pluginName(): string {
    return this._pluginName;
  }

  /**
   * Internal use only. Don't execute this method.
   * @hidden
   */
  public _privateInit(): void {
    if (!this._isReady) {
      this._isReady = true;
      this._queue._trigger("insert_at");
    }
  }

  public cancel(onComplete?: (error?: any) => void): Promise<void> {

    return new Promise((resolve: () => void, reject: (error: any) => void): void => {
      this.exec({
        args: [{
          targetId: this.id,
        }],
        context: this,
        methodName: "onDisconnect_cancel",
        pluginName: this.pluginName,
      })
      .then((): void => {
        resolve();
        if (typeof onComplete === "function") {
          onComplete();
        }
      })
      .catch((error: any) => {
        reject(error);
        if (typeof onComplete === "function") {
          onComplete(error);
        }
      });
    });
  }

  public remove(onComplete?: (error?: any) => void): Promise<void> {

    return new Promise((resolve: () => void, reject: (error: any) => void): void => {
      this.exec({
        args: [{
          targetId: this.id,
        }],
        context: this,
        methodName: "onDisconnect_remove",
        pluginName: this.pluginName,
      })
      .then((): void => {
        resolve();
        if (typeof onComplete === "function") {
          onComplete();
        }
      })
      .catch((error: any) => {
        reject(error);
        if (typeof onComplete === "function") {
          onComplete(error);
        }
      });
    });
  }

  public set(value: any, onComplete?: (error?: any) => void): Promise<void> {

    return new Promise((resolve: () => void, reject: (error: any) => void): void => {
      this.exec({
        args: [{
          targetId: this.id,
          value: LZString.compressToBase64(JSON.stringify(value)),
        }],
        context: this,
        methodName: "onDisconnect_set",
        pluginName: this.pluginName,
      })
      .then((): void => {
        resolve();
        if (typeof onComplete === "function") {
          onComplete();
        }
      })
      .catch((error: any) => {
        reject(error);
        if (typeof onComplete === "function") {
          onComplete(error);
        }
      });
    });
  }

  public setWithPriority(value: any, property: string | number, onComplete?: (error?: any) => void): Promise<void> {

    return new Promise((resolve: () => void, reject: (error: any) => void): void => {
      this.exec({
        args: [{
          priority: LZString.compressToBase64(JSON.stringify(property)),
          targetId: this.id,
          value: LZString.compressToBase64(JSON.stringify(value)),
        }],
        context: this,
        methodName: "onDisconnect_setWithPriority",
        pluginName: this.pluginName,
      })
      .then((): void => {
        resolve();
        if (typeof onComplete === "function") {
          onComplete();
        }
      })
      .catch((error: any) => {
        reject(error);
        if (typeof onComplete === "function") {
          onComplete(error);
        }
      });
    });
  }

  public update(values: any, onComplete?: (error?: any) => void): Promise<void> {

    if (!values || typeof values !== "object" || Array.isArray(values)) {
      throw new Error("values must be key-value object");
    }

    return new Promise((resolve: () => void, reject: (error: any) => void): void => {
      this.exec({
        args: [{
          targetId: this.id,
          values: LZString.compressToBase64(JSON.stringify(values)),
        }],
        context: this,
        methodName: "onDisconnect_update",
        pluginName: this.pluginName,
      })
      .then((): void => {
        resolve();
        if (typeof onComplete === "function") {
          onComplete.call(this);
        }
      })
      .catch((error: any): void => {
        reject(error);
        if (typeof onComplete === "function") {
          onComplete.call(this, error);
        }
      });
    });
  }

  private exec(params: IExecCmdParams): Promise<any> {
    return new Promise((resolve: (result: any) => void, reject: (error: any) => void) => {
      this._queue._push({
        args: params.args,
        context: params.context,
        execOptions: params.execOptions,
        methodName: params.methodName,
        pluginName: params.pluginName,
        reject,
        resolve,
      });
    });
  }

}
