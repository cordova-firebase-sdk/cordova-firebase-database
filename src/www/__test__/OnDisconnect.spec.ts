import { LZString } from "cordova-firebase-core/index";
import { OnDisconnect } from "../OnDisconnect";
import { execCmd } from "../__mocks__/CommandQueue";

describe("[OnDisconnect]", () => {
  beforeEach(() => {
    // Clear all instances and calls to constructor and all methods:
    execCmd.mockClear();
  });

  describe("internal tests", () => {
    it("should have correct properties.", () => {
      const _: OnDisconnect = new OnDisconnect("pluginName");
      expect(_.id.includes("_OnDisconnect")).toBe(true);
      expect(_.pluginName).toEqual("pluginName");
    });

    it("methods should work even if native side is not ready.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");

      const args1 = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify("test")),
      }];
      const args2 = [{
        targetId: _.id,
      }];

      _.set("test").then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("onDisconnect_set");
        expect(params.args).toEqual(args1);
      });
      _.cancel().then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[1][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_cancel");
        expect(params.args).toEqual(args2);
        done();
      });

      setTimeout(() => {
        _._privateInit();
      }, 10);
    });
  });

  describe("cancel()", () => {
    it("should involve native code with correct parameters.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.cancel().then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_cancel");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should work with onComplete.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.cancel(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_cancel");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should catch error of native side.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.cancel().catch((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_cancel");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });

    it("should catch error of native side with onComplete.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      try {
      _.cancel((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_cancel");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
      } catch(e) {
        console.error(e);
      }
    });
  });

  describe("remove()", () => {
    it("should involve native code with correct parameters.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.remove().then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("onDisconnect_remove");
        expect(params.args).toEqual(args);
        done();
      });
    });
    it("should work with onComplete.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.remove(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_remove");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should catch error of native side.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.remove().catch((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_remove");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });

    it("should catch error of native side with onComplete.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const args = [{
        targetId: _.id,
      }];

      _.remove((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_remove");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });
  });

  describe("set()", () => {
    it("should involve native code with correct parameters.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";

      const args = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.set(value).then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("onDisconnect_set");
        expect(params.args).toEqual(args);
        done();
      });
    });
    it("should work with onComplete.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";

      const args = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.set(value, () => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_set");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should catch error of native side.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";

      const args = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.set(value).catch((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_set");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });

    it("should catch error of native side with onComplete.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";

      const args = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.set(value, (error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_set");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });
  });

  describe("setWithPriority()", () => {
    it("should involve native code with correct parameters.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";
      const priority: string = "normal";

      const args = [{
        targetId: _.id,
        priority: LZString.compressToBase64(JSON.stringify(priority)),
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.setWithPriority(value, priority).then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("onDisconnect_setWithPriority");
        expect(params.args).toEqual(args);
        done();
      });
    });
    it("should work with onComplete.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";
      const priority: string = "normal";

      const args = [{
        targetId: _.id,
        priority: LZString.compressToBase64(JSON.stringify(priority)),
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.setWithPriority(value, priority, () => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_setWithPriority");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should catch error of native side.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";
      const priority: string = "normal";

      const args = [{
        targetId: _.id,
        priority: LZString.compressToBase64(JSON.stringify(priority)),
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.setWithPriority(value, priority).catch((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_setWithPriority");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });

    it("should catch error of native side with onComplete.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world";
      const priority: string = "normal";

      const args = [{
        targetId: _.id,
        priority: LZString.compressToBase64(JSON.stringify(priority)),
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.setWithPriority(value, priority, (error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_setWithPriority");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });
  });

  describe("update()", () => {
    it("should involve native code with correct parameters.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: object = {
        hello: "world",
      };

      const args = [{
        targetId: _.id,
        values: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.update(value).then(() => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("onDisconnect_update");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should work with onComplete.", (done) => {
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: object = {
        hello: "world",
      };

      const args = [{
        targetId: _.id,
        values: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.update(value, () => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_update");
        expect(params.args).toEqual(args);
        done();
      });
    });

    it("should catch error of native side.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: object = {
        hello: "world",
      };

      const args = [{
        targetId: _.id,
        values: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.update(value).catch((error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_update");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });

    it("should catch error of native side with onComplete.", (done) => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: object = {
        hello: "world",
      };

      const args = [{
        targetId: _.id,
        values: LZString.compressToBase64(JSON.stringify(value)),
      }];

      _.update(value, (error: Error): void => {
        expect(_.id.includes("_OnDisconnect")).toBe(true);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.pluginName).toEqual("dummyPluginName");
        expect(params.methodName).toEqual("onDisconnect_update");
        expect(error).toEqual(new Error("Error!"));
        done();
      });
    });
    it("should raise Error if value is not key-value paired object.", () => {
      execCmd.mockImplementationOnce((params: any): Promise<any> => {
        return Promise.reject(new Error("Error!"));
      });
      const _: OnDisconnect = new OnDisconnect("dummyPluginName");
      _._privateInit();

      const value: string = "hello world"

      const args = [{
        targetId: _.id,
        value: LZString.compressToBase64(JSON.stringify(value)),
      }];

      expect(() => {
        _.update(value);
      }).toThrowErrorMatchingSnapshot()
    });
  });
});
