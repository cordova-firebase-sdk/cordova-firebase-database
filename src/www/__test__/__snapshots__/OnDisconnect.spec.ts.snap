// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[OnDisconnect] update() should raise Error if value is not key-value paired object. 1`] = `"values must be key-value object"`;
